src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -726,16 +737,9 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
             !runPaths = Paths.runPath root (uniqueToRunNumber n)
         traceWith tr $ AtLevel ln $
           TraceNewMerge (V.map Run.size rs) (runNumber runPaths) caching alloc mergePolicy mergeLevel
-        -- TODO: There currently is a resource management bug that happens if an
-        -- exception occurs after calling MR.new. In this case, all changes roll
-        -- back, so some of the runs in rs will live in the Levels structure at
-        -- their original places again. However, we passed their references to
-        -- the MergingRun, which gets aborted, releasing the run references.
-        -- Instead of passing the original references into newMerge, we have to
-        -- duplicate the ones that previously existed in the level and then
-        -- freeTemp the original ones. This way, on the happy path the result is
-        -- the same, but if an exception occurs, the original references do not
-        -- get released.
+        -- The runs will end up inside the merging run, with fresh references.
+        -- The original references can be released (but only on the happy path).
+        V.forM_ rs $ \r -> freeTemp reg (releaseRef r)
         mr <- allocateTemp reg
           (MR.new hfs hbio resolve caching alloc mergeLevel mergePolicy runPaths rs)
           releaseRef
```

Comment:
It's not important for correctness, but it might be nicer for uninitiated readers if these releases happen after the new merging run is created -- that's closer to how the code would look if we didn't have a temp registry: can't release resources before we use them

Reply:
done